File: ios/UmbraCore/Sources/UmbraCore/Net/DnsMessage.swift
import Foundation

/// Minimal DNS: read the first question, and build a sinkholed response for a
/// blocked domain (A -> 0.0.0.0, AAAA -> ::, else NXDOMAIN). Mirrors the Android
/// DnsMessage; the algorithm is cross-checked against the Python mirror.
public struct DnsQuery {
    public static let typeA = 1
    public static let typeAAAA = 28
    private static let classIN = 1
    private static let rcodeNXDOMAIN = 3
    private static let questionStart = 12

    public let id: Int
    public let flags: Int
    public let qName: String
    public let qType: Int
    public let raw: [UInt8]
    public let questionEnd: Int

    public static func parse(_ payload: [UInt8]) -> DnsQuery? {
        if payload.count < 12 { return nil }
        let id = u16(payload, 0)
        let flags = u16(payload, 2)
        let qd = u16(payload, 4)
        if qd < 1 { return nil }

        var name = ""
        var i = questionStart
        while i < payload.count {
            let len = Int(payload[i])
            if len == 0 { i += 1; break }
            if (len & 0xC0) != 0 { return nil }
            if i + 1 + len > payload.count { return nil }
            if !name.isEmpty { name += "." }
            for j in 0..<len { name.append(Character(UnicodeScalar(payload[i + 1 + j]))) }
            i += 1 + len
        }
        if i + 4 > payload.count { return nil }
        let qType = u16(payload, i)
        i += 4
        return DnsQuery(id: id, flags: flags, qName: name.lowercased(),
                        qType: qType, raw: payload, questionEnd: i)
    }

    /// A response that sinkholes this query.
    public func buildBlockedResponse() -> [UInt8] {
        let questionLen = questionEnd - DnsQuery.questionStart
        let rdata: [UInt8]? = qType == DnsQuery.typeA ? [UInt8](repeating: 0, count: 4)
            : (qType == DnsQuery.typeAAAA ? [UInt8](repeating: 0, count: 16) : nil)
        let hasAnswer = rdata != nil
        let anc = hasAnswer ? 1 : 0
        let answerLen = hasAnswer ? 12 + rdata!.count : 0

        var out = [UInt8](repeating: 0, count: 12 + questionLen + answerLen)
        DnsQuery.put16(&out, 0, id)
        let rcode = hasAnswer ? 0 : DnsQuery.rcodeNXDOMAIN
        let rd = flags & 0x0100
        DnsQuery.put16(&out, 2, 0x8000 | rd | 0x0080 | rcode)
        DnsQuery.put16(&out, 4, 1)
        DnsQuery.put16(&out, 6, anc)
        DnsQuery.put16(&out, 8, 0); DnsQuery.put16(&out, 10, 0)

        for k in 0..<questionLen { out[12 + k] = raw[DnsQuery.questionStart + k] }

        if hasAnswer, let rdata = rdata {
            var o = 12 + questionLen
            DnsQuery.put16(&out, o, 0xC000 | DnsQuery.questionStart); o += 2
            DnsQuery.put16(&out, o, qType); o += 2
            DnsQuery.put16(&out, o, DnsQuery.classIN); o += 2
            DnsQuery.put16(&out, o, 0); DnsQuery.put16(&out, o + 2, 60); o += 4
            DnsQuery.put16(&out, o, rdata.count); o += 2
            for (k, b) in rdata.enumerated() { out[o + k] = b }
        }
        return out
    }

    private static func u16(_ b: [UInt8], _ o: Int) -> Int { (Int(b[o]) << 8) | Int(b[o + 1]) }
    private static func put16(_ b: inout [UInt8], _ o: Int, _ v: Int) {
        b[o] = UInt8((v >> 8) & 0xFF); b[o + 1] = UInt8(v & 0xFF)
    }
}

File: android/app/src/main/java/com/forrestlasiter/umbra/net/dns/DnsMessage.kt
package com.forrestlasiter.umbra.net.dns

/**
 * Minimal DNS wire parsing: read the first question (name + type) from a query,
 * and build a "sinkholed" response for a blocked domain. We answer A with
 * 0.0.0.0 (exactly what the Linux /etc/hosts sinkhole does), AAAA with ::, and
 * anything else with NXDOMAIN. Only the question is parsed — no need to walk
 * answer/authority sections or follow compression pointers (a question name is
 * never compressed).
 */
data class DnsQuery(
    val id: Int,
    val flags: Int,
    val qName: String,     // lower-case, no trailing dot, e.g. "graph.facebook.com"
    val qType: Int,
    val raw: ByteArray,
    val questionEnd: Int,  // offset just past qname+qtype+qclass
) {
    companion object {
        const val TYPE_A = 1
        const val TYPE_AAAA = 28
        private const val CLASS_IN = 1
        private const val RCODE_NXDOMAIN = 3
        private const val QUESTION_START = 12

        fun parse(payload: ByteArray): DnsQuery? {
            if (payload.size < 12) return null
            val id = u16(payload, 0)
            val flags = u16(payload, 2)
            val qdcount = u16(payload, 4)
            if (qdcount < 1) return null

            val sb = StringBuilder()
            var i = QUESTION_START
            while (i < payload.size) {
                val len = payload[i].toInt() and 0xFF
                if (len == 0) { i += 1; break }
                if (len and 0xC0 != 0) return null           // compression not expected in a question
                if (i + 1 + len > payload.size) return null
                if (sb.isNotEmpty()) sb.append('.')
                for (j in 0 until len) sb.append((payload[i + 1 + j].toInt() and 0xFF).toChar())
                i += 1 + len
            }
            if (i + 4 > payload.size) return null
            val qType = u16(payload, i)
            i += 4                                            // skip qtype + qclass
            return DnsQuery(id, flags, sb.toString().lowercase(), qType, payload, i)
        }

        private fun u16(b: ByteArray, o: Int) =
            ((b[o].toInt() and 0xFF) shl 8) or (b[o + 1].toInt() and 0xFF)
    }

    /** A response that sinkholes this query (drops it into a black hole). */
    fun buildBlockedResponse(): ByteArray {
        val questionLen = questionEnd - QUESTION_START
        val rdata: ByteArray? = when (qType) {
            TYPE_A -> ByteArray(4)        // 0.0.0.0
            TYPE_AAAA -> ByteArray(16)    // ::
            else -> null                  // -> NXDOMAIN, no answer record
        }
        val hasAnswer = rdata != null
        val ancount = if (hasAnswer) 1 else 0
        val answerLen = if (hasAnswer) 12 + rdata!!.size else 0  // ptr(2)+type(2)+class(2)+ttl(4)+rdlen(2)+rdata

        val out = ByteArray(12 + questionLen + answerLen)
        // Header
        put16(out, 0, id)
        // QR=1, opcode+RD echoed from query, RA=1, rcode
        val rcode = if (hasAnswer) 0 else RCODE_NXDOMAIN
        val rd = flags and 0x0100                              // preserve Recursion Desired
        put16(out, 2, 0x8000 or rd or 0x0080 or rcode)         // QR | RD? | RA | rcode
        put16(out, 4, 1)                                       // qdcount
        put16(out, 6, ancount)                                 // ancount
        put16(out, 8, 0); put16(out, 10, 0)                    // ns/ar count

        // Echo the question verbatim.
        System.arraycopy(raw, QUESTION_START, out, 12, questionLen)

        // Answer (if any), name compressed to the question at offset 12.
        if (hasAnswer) {
            var o = 12 + questionLen
            put16(out, o, 0xC000 or QUESTION_START); o += 2    // name pointer -> 0x000C
            put16(out, o, qType); o += 2
            put16(out, o, CLASS_IN); o += 2
            put16(out, o, 0); put16(out, o + 2, 60); o += 4    // TTL = 60s
            put16(out, o, rdata!!.size); o += 2
            System.arraycopy(rdata, 0, out, o, rdata.size)
        }
        return out
    }

    private fun put16(b: ByteArray, o: Int, v: Int) {
        b[o] = ((v ushr 8) and 0xFF).toByte()
        b[o + 1] = (v and 0xFF).toByte()
    }
}
